rfcs/0000-rlp-v2.md

```diff
@@ -0,0 +1,740 @@
+      <td>
+        <ul>
+          <li>Less generic naming</li>
+        </ul>
+      </td>
+    </tr>
+    <tr>
+      <td><code>spec.rateLimits.limits.maxValue</code></td>
+      <td><code>spec.limits.rates.limit</code></td>
+      <td>
+        <ul>
+          <li>More generic naming</li>
+        </ul>
+      </td>
+    </tr>
+    <tr>
+      <td>Fixed implicit increment of 1</td>
+      <td><code>spec.limits.increment: {}</code></td>
+      <td>
+        <ul>
+          <li>Flexibility to change the limits without reseting the counters</li>
+          <li>Placeholder for more complex RL systems in the future – e.g. dynamic-RL, %, 𝑓(), etc</li>
+          <li>Migration path from other RL solutions, e.g. 3scale (?)</li>
+        </ul>
+      </td>
+    </tr>
+  </tbody>
+</table>
+
+## Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+By completely dropping out the `configurations` field from the RLP, [composing the RL descriptor actions](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/rate_limit_filter#composing-actions) is now done based essentially on the selectors listed in the `when` conditions and the `counters`, plus artificial generic conditions used to match the specific route rule (when `matches` is specified).
+
+### Well-known selectors
+
+Each selector is a direct reference to a path within a well-known data structure that holds the `context` (L4 and L7 data of the original request handled by the proxy), as well as occasionally the `auth` data (dynamic metadata exported by the external authorization filter and injected by the proxy into the rate-limit filter).
+
+The well-known data structure for building RL descriptor actions resembles Authorino's ["Authorization JSON"](https://github.com/Kuadrant/authorino/blob/main/docs/architecture.md#the-authorization-json), whose `context` component consists of Envoy's [`AttributeContext`](https://pkg.go.dev/github.com/envoyproxy/go-control-plane/envoy/service/auth/v3?utm_source=gopls#AttributeContext) type of the external authorization API, marshalled as JSON. Compared to the more generic [`RateLimitRequest`](https://pkg.go.dev/github.com/envoyproxy/go-control-plane@v0.11.0/envoy/service/ratelimit/v3#RateLimitRequest) struct, the `AttributeContext` provides a more structured and arguibly more intuitive relation between the data sources for the RL descriptors actions and their corresponding key names through which the values are referred within the RLP, in a context of serving predominantly for HTTP-based APIs.
+
+To keep compatibility with the [Envoy Rate Limit protocol](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/rate_limit_filter), the well-known data structure can optionally be extended with the `RateLimitRequest`, thus resulting in the following final structure.
+
+```yaml
+context: # Envoy's Ext-Authz `CheckRequest.AttributeContext` type
```

Comment:
The wasm module has access to the Envoy context. Passing a "path", you can get values.

Some path examples used now:

[source_cluster](https://github.com/Kuadrant/wasm-shim/blob/03d00094d74195df25127dc69656cd6edd5af197/src/filter/http_context.rs#L149): 
```rust
match self.get_property(vec!["connection", "requested_server_name"]) {
```

[metadata](https://github.com/Kuadrant/wasm-shim/blob/03d00094d74195df25127dc69656cd6edd5af197/src/filter/http_context.rs#L250):
```rust
self.get_property(vec!["metadata", "filter_metadata", key]) 
```

Reply:
The wasm module is an implementation detail that belongs to the data plane. This is about building the configuration in the control plane.

I'm proposing a well-known data structure that specifies to users what attributes of a request can be referred in a policy and how. Envoy's Ext-Authz `CheckRequest.AttributeContext` type was proposed as a base for such well-known data structure because not only it is suitable for this, providing enough context for any TCP data exchange, but also because it would allow us to have an homogeneous API compared to the `AuthPolicy` without having to make this RFC about the `AuthPolicy` as well just yet.